File: src/Serialization/HybridRow.Tests.Unit/SystemSchemaUnitTests.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

namespace Microsoft.Azure.Cosmos.Serialization.HybridRow.Tests.Unit
{
    using System.IO;
    using Microsoft.Azure.Cosmos.Serialization.HybridRow.Layouts;
    using Microsoft.Azure.Cosmos.Serialization.HybridRow.Schemas;
    using Microsoft.VisualStudio.TestTools.UnitTesting;
    using Microsoft.VisualStudio.TestTools.UnitTesting.Logging;

    [TestClass]
    public class SystemSchemaUnitTests
    {
        private const string CrossVersioningSchema = @"TestData\CrossVersioningSchema.json";
        private const string CustomerSchema = @"TestData\CustomerSchema.json";
        private const string NullableSchema = @"TestData\NullableSchema.json";
        private const string ReaderSchema = @"TestData\ReaderSchema.json";
        private const string SchemaHashCoverageSchema = @"TestData\SchemaHashCoverageSchema.json";
        private const string CoverageSchema = @"TestData\CoverageSchema.json";
        private const string BatchApiSchema = @"TestData\BatchApiSchema.json";
        private const string TaggedApiSchema = @"TestData\TaggedApiSchema.json";
        private const string PerfCounterSchema = @"TestData\PerfCounterSchema.json";
        private const string TagSchema = @"TestData\TagSchema.json";
        private const string MovieSchema = @"TestData\MovieSchema.json";
        private const string SchemaTodoSchema = @"TestData\TodoSchema.json";

        [TestMethod]
        [Owner("jthunter")]
        public void LoadGeneratedHrSchema()
        {
            Layout layout = SchemasHrSchema.LayoutResolver.Resolve((SchemaId)SchemaOptionsHybridRowSerializer.SchemaId);
            Assert.AreEqual((SchemaId)SchemaOptionsHybridRowSerializer.SchemaId, layout.SchemaId);
        }

        [TestMethod]
        [Owner("jthunter")]
        public void LoadTest()
        {
            LayoutResolver resolver = SystemSchema.LayoutResolver;

            SchemaId[] systemSchemas =
            {
                SystemSchema.EmptySchemaId,
                (SchemaId)SegmentHybridRowSerializer.SchemaId,
                (SchemaId)RecordHybridRowSerializer.SchemaId,
                (SchemaId)NamespaceHybridRowSerializer.SchemaId,
                (SchemaId)SchemaHybridRowSerializer.SchemaId,
                (SchemaId)SchemaOptionsHybridRowSerializer.SchemaId,
                (SchemaId)PartitionKeyHybridRowSerializer.SchemaId,
                (SchemaId)PrimarySortKeyHybridRowSerializer.SchemaId,
                (SchemaId)StaticKeyHybridRowSerializer.SchemaId,
                (SchemaId)PropertyHybridRowSerializer.SchemaId,
                (SchemaId)PropertyTypeHybridRowSerializer.SchemaId,
                (SchemaId)PrimitivePropertyTypeHybridRowSerializer.SchemaId,
                (SchemaId)ScopePropertyTypeHybridRowSerializer.SchemaId,
                (SchemaId)ArrayPropertyTypeHybridRowSerializer.SchemaId,
                (SchemaId)ObjectPropertyTypeHybridRowSerializer.SchemaId,
                (SchemaId)SetPropertyTypeHybridRowSerializer.SchemaId,
                (SchemaId)MapPropertyTypeHybridRowSerializer.SchemaId,
                (SchemaId)TuplePropertyTypeHybridRowSerializer.SchemaId,
                (SchemaId)TaggedPropertyTypeHybridRowSerializer.SchemaId,
            };

            // Make sure all system schemas are loadable.
            foreach (SchemaId id in systemSchemas)
            {
                Layout l = resolver.Resolve(id);
                Assert.AreEqual(id, l.SchemaId);
            }

            // Make sure all system schema ids are unique.
            foreach (SchemaId id in systemSchemas)
            {
                int count = 0;
                foreach (SchemaId other in systemSchemas)
                {
                    if (other == id)
                    {
                        count++;
                    }
                }

                Assert.AreEqual(1, count);
            }
        }

        [TestMethod]
        [Owner("jthunter")]
        public void SerializeSystemNamespaceTest()
        {
            Namespace ns1 = SystemSchema.GetNamespace();
            SystemSchemaUnitTests.SerializerRoundtripNamespace(ns1);
        }

        [TestMethod]
        [Owner("jthunter")]
        [DeploymentItem(SystemSchemaUnitTests.CrossVersioningSchema, "TestData")]
        [DeploymentItem(SystemSchemaUnitTests.CustomerSchema, "TestData")]
        [DeploymentItem(SystemSchemaUnitTests.NullableSchema, "TestData")]
        [DeploymentItem(SystemSchemaUnitTests.ReaderSchema, "TestData")]
        [DeploymentItem(SystemSchemaUnitTests.SchemaHashCoverageSchema, "TestData")]
        [DeploymentItem(SystemSchemaUnitTests.CoverageSchema, "TestData")]
        [DeploymentItem(SystemSchemaUnitTests.BatchApiSchema, "TestData")]
        [DeploymentItem(SystemSchemaUnitTests.TaggedApiSchema, "TestData")]
        [DeploymentItem(SystemSchemaUnitTests.PerfCounterSchema, "TestData")]
        [DeploymentItem(SystemSchemaUnitTests.TagSchema, "TestData")]
        [DeploymentItem(SystemSchemaUnitTests.MovieSchema, "TestData")]
        [DeploymentItem(SystemSchemaUnitTests.SchemaTodoSchema, "TestData")]
        public void SerializeTestSchemasTest()
        {
            string[] files =
            {
                SystemSchemaUnitTests.CrossVersioningSchema,
                SystemSchemaUnitTests.CustomerSchema,
                SystemSchemaUnitTests.NullableSchema,
                SystemSchemaUnitTests.ReaderSchema,
                SystemSchemaUnitTests.SchemaHashCoverageSchema,
                SystemSchemaUnitTests.CoverageSchema,
                SystemSchemaUnitTests.BatchApiSchema,
                SystemSchemaUnitTests.TaggedApiSchema,
                SystemSchemaUnitTests.PerfCounterSchema,
                SystemSchemaUnitTests.TagSchema,
                SystemSchemaUnitTests.MovieSchema,
                SystemSchemaUnitTests.SchemaTodoSchema,
            };

            foreach (string filename in files)
            {
                Logger.LogMessage("Filename: {0}", filename);
                string json = File.ReadAllText(filename);
                Namespace ns1 = Namespace.Parse(json);
                SystemSchemaUnitTests.SerializerRoundtripNamespace(ns1);
            }
        }

        private static void SerializerRoundtripNamespace(Namespace ns1)
        {
            Layout layout = SystemSchema.LayoutResolver.Resolve((SchemaId)NamespaceHybridRowSerializer.SchemaId);

            RowBuffer row = new RowBuffer(0);
            row.InitLayout(HybridRowVersion.V1, layout, SystemSchema.LayoutResolver);

            // Write the whole namespace to a row.
            Result r = ns1.Write(ref row);
            ResultAssert.IsSuccess(r);

            // Read the namespace back.
            r = Namespace.Read(ref row, out Namespace ns2);
            ResultAssert.IsSuccess(r);

            // Compare the materialized row with the original in-memory object model.
            Assert.AreEqual(ns1.Version, ns2.Version);
            Assert.AreEqual(ns1.Name, ns2.Name);
            Assert.AreEqual(ns1.Comment, ns2.Comment);
            Assert.AreEqual(ns1.CppNamespace, ns2.CppNamespace);
            Assert.AreEqual(ns1.Schemas.Count, ns2.Schemas.Count);
            for (int i = 0; i < ns1.Schemas.Count; i++)
            {
                Schema s1 = ns1.Schemas[i];
                Schema s2 = ns2.Schemas[i];
                SystemSchemaUnitTests.AssertEqual(s1, s2);
            }
            Assert.AreEqual(ns1.Enums.Count, ns2.Enums.Count);
            for (int i = 0; i < ns1.Enums.Count; i++)
            {
                EnumSchema s1 = ns1.Enums[i];
                EnumSchema s2 = ns2.Enums[i];
                SystemSchemaUnitTests.AssertEqual(s1, s2);
            }
        }

        private static void AssertEqual(Schema s1, Schema s2)
        {
            Assert.AreEqual(s1.Version, s2.Version);
            Assert.AreEqual(s1.Type, s2.Type);
            Assert.AreEqual(s1.SchemaId, s2.SchemaId);
            Assert.AreEqual(s1.Name, s2.Name);
            Assert.AreEqual(s1.Comment, s2.Comment);
            Assert.AreEqual(s1.BaseName, s2.BaseName);
            Assert.AreEqual(s1.BaseSchemaId, s2.BaseSchemaId);

            if (s1.Options == null)
            {
                Assert.IsNull(s2.Options);
            }
            else
            {
                Assert.IsNotNull(s2.Options);
                Assert.AreEqual(s1.Options.DisallowUnschematized, s2.Options.DisallowUnschematized);
                Assert.AreEqual(s1.Options.EnablePropertyLevelTimestamp, s2.Options.EnablePropertyLevelTimestamp);
                Assert.AreEqual(s1.Options.DisableSystemPrefix, s2.Options.DisableSystemPrefix);
                Assert.AreEqual(s1.Options.Abstract, s2.Options.Abstract);
            }

            Assert.AreEqual(s1.PartitionKeys.Count, s2.PartitionKeys.Count);
            for (int i = 0; i < s1.PartitionKeys.Count; i++)
            {
                SystemSchemaUnitTests.AssertEqual(s1.PartitionKeys[i], s2.PartitionKeys[i]);
            }

            Assert.AreEqual(s1.PrimaryKeys.Count, s2.PrimaryKeys.Count);
            for (int i = 0; i < s1.PrimaryKeys.Count; i++)
            {
                SystemSchemaUnitTests.AssertEqual(s1.PrimaryKeys[i], s2.PrimaryKeys[i]);
            }

            Assert.AreEqual(s1.StaticKeys.Count, s2.StaticKeys.Count);
            for (int i = 0; i < s1.StaticKeys.Count; i++)
            {
                SystemSchemaUnitTests.AssertEqual(s1.StaticKeys[i], s2.StaticKeys[i]);
            }

            Assert.AreEqual(s1.Properties.Count, s2.Properties.Count);
            for (int i = 0; i < s1.Properties.Count; i++)
            {
                SystemSchemaUnitTests.AssertEqual(s1.Properties[i], s2.Properties[i]);
            }
        }

        private static void AssertEqual(PartitionKey i1, PartitionKey i2)
        {
            Assert.AreEqual(i1.Path, i2.Path);
        }

        private static void AssertEqual(PrimarySortKey i1, PrimarySortKey i2)
        {
            Assert.AreEqual(i1.Path, i2.Path);
            Assert.AreEqual(i1.Direction, i2.Direction);
        }

        private static void AssertEqual(StaticKey i1, StaticKey i2)
        {
            Assert.AreEqual(i1.Path, i2.Path);
        }

        private static void AssertEqual(Property i1, Property i2)
        {
            Assert.AreEqual(i1.Comment, i2.Comment);
            Assert.AreEqual(i1.Path, i2.Path);
            Assert.AreEqual(i1.ApiName, i2.ApiName);
            Assert.AreEqual(i1.AllowEmpty, i2.AllowEmpty);
            SystemSchemaUnitTests.AssertEqual(i1.PropertyType, i2.PropertyType);
        }

        private static void AssertEqual(PropertyType i1, PropertyType i2)
        {
            Assert.AreEqual(i1.ApiType, i2.ApiType);
            Assert.AreEqual(i1.Type, i2.Type);
            Assert.AreEqual(i1.Nullable, i2.Nullable);
            Assert.AreEqual(i1.GetType(), i2.GetType());

            switch (i1)
            {
                case PrimitivePropertyType p:
                    SystemSchemaUnitTests.AssertEqual(p, (PrimitivePropertyType)i2);
                    return;
                case ScopePropertyType p:
                    SystemSchemaUnitTests.AssertEqual(p, (ScopePropertyType)i2);
                    return;
                default:
                    Assert.Fail("Type is abstract.");
                    return;
            }
        }

        private static void AssertEqual(PrimitivePropertyType i1, PrimitivePropertyType i2)
        {
            Assert.AreEqual(i1.Length, i2.Length);
            Assert.AreEqual(i1.Storage, i2.Storage);
        }

        private static void AssertEqual(ScopePropertyType i1, ScopePropertyType i2)
        {
            Assert.AreEqual(i1.Immutable, i2.Immutable);

            switch (i1)
            {
                case ArrayPropertyType p:
                    SystemSchemaUnitTests.AssertEqual(p.Items, ((ArrayPropertyType)i2).Items);
                    break;
                case ObjectPropertyType p:
                {
                    ObjectPropertyType q = (ObjectPropertyType)i2;
                    Assert.AreEqual(p.Properties.Count, q.Properties.Count);
                    for (int i = 0; i < p.Properties.Count; i++)
                    {
                        SystemSchemaUnitTests.AssertEqual(p.Properties[i], q.Properties[i]);
                    }

                    break;
                }

                case SetPropertyType p:
                    SystemSchemaUnitTests.AssertEqual(p.Items, ((SetPropertyType)i2).Items);
                    break;
                case MapPropertyType p:
                    SystemSchemaUnitTests.AssertEqual(p.Keys, ((MapPropertyType)i2).Keys);
                    SystemSchemaUnitTests.AssertEqual(p.Values, ((MapPropertyType)i2).Values);
                    break;
                case TuplePropertyType p:
                {
                    TuplePropertyType q = (TuplePropertyType)i2;
                    Assert.AreEqual(p.Items.Count, q.Items.Count);
                    for (int i = 0; i < p.Items.Count; i++)
                    {
                        SystemSchemaUnitTests.AssertEqual(p.Items[i], q.Items[i]);
                    }

                    break;
                }

                case TaggedPropertyType p:
                {
                    TaggedPropertyType q = (TaggedPropertyType)i2;
                    Assert.AreEqual(p.Items.Count, q.Items.Count);
                    for (int i = 0; i < p.Items.Count; i++)
                    {
                        SystemSchemaUnitTests.AssertEqual(p.Items[i], q.Items[i]);
                    }

                    break;
                }

                case UdtPropertyType p:
                    Assert.AreEqual(p.Name, ((UdtPropertyType)i2).Name);
                    Assert.AreEqual(p.SchemaId, ((UdtPropertyType)i2).SchemaId);
                    break;
                default:
                    Assert.Fail("Type is abstract.");
                    return;
            }
        }

        private static void AssertEqual(EnumSchema s1, EnumSchema s2)
        {
            Assert.AreEqual(s1.Type, s2.Type);
            Assert.AreEqual(s1.ApiType, s2.ApiType);
            Assert.AreEqual(s1.Name, s2.Name);
            Assert.AreEqual(s1.Comment, s2.Comment);

            Assert.AreEqual(s1.Values.Count, s2.Values.Count);
            for (int i = 0; i < s1.Values.Count; i++)
            {
                SystemSchemaUnitTests.AssertEqual(s1.Values[i], s2.Values[i]);
            }
        }

        private static void AssertEqual(EnumValue s1, EnumValue s2)
        {
            Assert.AreEqual(s1.Value, s2.Value);
            Assert.AreEqual(s1.Name, s2.Name);
            Assert.AreEqual(s1.Comment, s2.Comment);
        }
    }
}

File: src/Serialization/HybridRow.Native.Tests.Unit/SystemSchemaUnitTests.cpp
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

#include "pch.h"
#include "CppUnitTest.h"
#include "CppUnitTestFramework.inl"
#include "ResultAssert.h"

// ReSharper disable CppClangTidyCppcoreguidelinesProTypeStaticCastDowncast
namespace cdb_hr_test
{
  using namespace std::literals;
  using Assert = Microsoft::VisualStudio::CppUnitTestFramework::Assert;

  TEST_CLASS(SystemSchemaUnitTests)
  {
  public:

  TEST_METHOD_WITH_OWNER(LoadGeneratedHrSchema, "jthunter")
    {
      const cdb_hr::Layout& layout = cdb_hr::SchemasHrSchema::GetLayoutResolver().Resolve(cdb_hr::SchemaOptionsHybridRowSerializer::Id);
      Assert::AreEqual(cdb_hr::SchemaOptionsHybridRowSerializer::Id, layout.GetSchemaId());
    }

  TEST_METHOD_WITH_OWNER(LoadTest, "jthunter")
    {
      const cdb_hr::LayoutResolver& resolver = cdb_hr::SchemasHrSchema::GetLayoutResolver();

      std::array systemSchemas{
        cdb_hr::SystemSchemaLiteral::EmptySchemaId,
        cdb_hr::SegmentHybridRowSerializer::Id,
        cdb_hr::RecordHybridRowSerializer::Id,
        cdb_hr::NamespaceHybridRowSerializer::Id,
        cdb_hr::SchemaHybridRowSerializer::Id,
        cdb_hr::SchemaOptionsHybridRowSerializer::Id,
        cdb_hr::PartitionKeyHybridRowSerializer::Id,
        cdb_hr::PrimarySortKeyHybridRowSerializer::Id,
        cdb_hr::StaticKeyHybridRowSerializer::Id,
        cdb_hr::PropertyHybridRowSerializer::Id,
        cdb_hr::PropertyTypeHybridRowSerializer::Id,
        cdb_hr::PrimitivePropertyTypeHybridRowSerializer::Id,
        cdb_hr::ScopePropertyTypeHybridRowSerializer::Id,
        cdb_hr::ArrayPropertyTypeHybridRowSerializer::Id,
        cdb_hr::ObjectPropertyTypeHybridRowSerializer::Id,
        cdb_hr::SetPropertyTypeHybridRowSerializer::Id,
        cdb_hr::MapPropertyTypeHybridRowSerializer::Id,
        cdb_hr::TuplePropertyTypeHybridRowSerializer::Id,
        cdb_hr::TaggedPropertyTypeHybridRowSerializer::Id,
      };

      // Make sure all system schemas are loadable.
      for (cdb_hr::SchemaId id : systemSchemas)
      {
        const cdb_hr::Layout& l = resolver.Resolve(id);
        Assert::AreEqual(id, l.GetSchemaId());
      }

      // Make sure all system schema ids are unique.
      for (cdb_hr::SchemaId id : systemSchemas)
      {
        int count = 0;
        for (cdb_hr::SchemaId other : systemSchemas)
        {
          if (other == id)
          {
            count++;
          }
        }

        Assert::AreEqual(1, count);
      }
    }

    static void AssertEqual(const cdb_hr::PartitionKey& i1, const cdb_hr::PartitionKey& i2)
    {
      Assert::AreEqual(i1.GetPath(), i2.GetPath());
    }

    static void AssertEqual(const cdb_hr::PrimarySortKey& i1, const cdb_hr::PrimarySortKey& i2)
    {
      Assert::AreEqual(i1.GetPath(), i2.GetPath());
      Assert::AreEqual(i1.GetDirection(), i2.GetDirection());
    }

    static void AssertEqual(const cdb_hr::StaticKey& i1, const cdb_hr::StaticKey& i2)
    {
      Assert::AreEqual(i1.GetPath(), i2.GetPath());
    }

    static void AssertEqual(const cdb_hr::Property& i1, const cdb_hr::Property& i2)
    {
      Assert::AreEqual(i1.GetComment(), i2.GetComment());
      Assert::AreEqual(i1.GetPath(), i2.GetPath());
      Assert::AreEqual(i1.GetApiName(), i2.GetApiName());
      Assert::AreEqual(i1.GetAllowEmpty(), i2.GetAllowEmpty());
      if (i1.GetPropertyType().has_value())
      {
        Assert::IsTrue(i2.GetPropertyType().has_value());
        AssertEqual(i1.GetPropertyType().value(), i2.GetPropertyType().value());
      }
      else
      {
        Assert::IsFalse(i2.GetPropertyType().has_value());
      }
    }

    static void AssertEqual(const cdb_hr::PropertyType& i1, const cdb_hr::PropertyType& i2)
    {
      Assert::AreEqual(i1.GetApiType(), i2.GetApiType());
      Assert::AreEqual(i1.GetType(), i2.GetType());
      Assert::AreEqual(i1.GetNullable(), i2.GetNullable());

      Assert::AreEqual(i1.GetRuntimeSchemaId(), i2.GetRuntimeSchemaId());
      switch (i1.GetRuntimeSchemaId().Id())
      {
      case cdb_hr::PrimitivePropertyTypeHybridRowSerializer::Id.Id():
      {
        AssertEqual(static_cast<const cdb_hr::PrimitivePropertyType&>(i1), static_cast<const cdb_hr::PrimitivePropertyType&>(i2));
        return;
      }

      case cdb_hr::ArrayPropertyTypeHybridRowSerializer::Id.Id():
      case cdb_hr::ObjectPropertyTypeHybridRowSerializer::Id.Id():
      case cdb_hr::UdtPropertyTypeHybridRowSerializer::Id.Id():
      case cdb_hr::SetPropertyTypeHybridRowSerializer::Id.Id():
      case cdb_hr::MapPropertyTypeHybridRowSerializer::Id.Id():
      case cdb_hr::TuplePropertyTypeHybridRowSerializer::Id.Id():
      case cdb_hr::TaggedPropertyTypeHybridRowSerializer::Id.Id():
      case cdb_hr::ScopePropertyTypeHybridRowSerializer::Id.Id():
      {
        AssertEqual(static_cast<const cdb_hr::ScopePropertyType&>(i1), static_cast<const cdb_hr::ScopePropertyType&>(i2));
        return;
      }

      default:
        Assert::Fail(L"Type is abstract.");
      }
    }

    static void AssertEqual(const cdb_hr::PrimitivePropertyType& i1, const cdb_hr::PrimitivePropertyType& i2)
    {
      Assert::AreEqual(i1.GetLength(), i2.GetLength());
      Assert::AreEqual(i1.GetStorage(), i2.GetStorage());
    }

    static void AssertEqual(const cdb_hr::ScopePropertyType& i1, const cdb_hr::ScopePropertyType& i2)
    {
      Assert::AreEqual(i1.GetImmutable(), i2.GetImmutable());

      switch (i1.GetRuntimeSchemaId().Id())
      {
      case cdb_hr::ArrayPropertyTypeHybridRowSerializer::Id.Id():
      {
        const cdb_hr::ArrayPropertyType& p = static_cast<const cdb_hr::ArrayPropertyType&>(i1);
        const cdb_hr::ArrayPropertyType& q = static_cast<const cdb_hr::ArrayPropertyType&>(i2);
        Assert::AreEqual(p.GetItems().has_value(), q.GetItems().has_value());
        AssertEqual(p.GetItems().value(), q.GetItems().value());
        break;
      }
      case cdb_hr::ObjectPropertyTypeHybridRowSerializer::Id.Id():
      {
        const cdb_hr::ObjectPropertyType& p = static_cast<const cdb_hr::ObjectPropertyType&>(i1);
        const cdb_hr::ObjectPropertyType& q = static_cast<const cdb_hr::ObjectPropertyType&>(i2);
        Assert::AreEqual(p.GetProperties().size(), q.GetProperties().size());
        for (size_t i = 0; i < p.GetProperties().size(); i++)
        {
          AssertEqual(*p.GetProperties()[i], *q.GetProperties()[i]);
        }

        break;
      }
      case cdb_hr::SetPropertyTypeHybridRowSerializer::Id.Id():
      {
        const cdb_hr::SetPropertyType& p = static_cast<const cdb_hr::SetPropertyType&>(i1);
        const cdb_hr::SetPropertyType& q = static_cast<const cdb_hr::SetPropertyType&>(i2);
        Assert::AreEqual(p.GetItems().has_value(), q.GetItems().has_value());
        AssertEqual(p.GetItems().value(), q.GetItems().value());
        break;
      }
      case cdb_hr::MapPropertyTypeHybridRowSerializer::Id.Id():
      {
        const cdb_hr::MapPropertyType& p = static_cast<const cdb_hr::MapPropertyType&>(i1);
        const cdb_hr::MapPropertyType& q = static_cast<const cdb_hr::MapPropertyType&>(i2);
        Assert::AreEqual(p.GetKeys().has_value(), q.GetKeys().has_value());
        Assert::AreEqual(p.GetValues().has_value(), q.GetValues().has_value());
        AssertEqual(p.GetKeys().value(), q.GetKeys().value());
        AssertEqual(p.GetValues().value(), q.GetValues().value());
        break;
      }
      case cdb_hr::TuplePropertyTypeHybridRowSerializer::Id.Id():
      {
        const cdb_hr::TuplePropertyType& p = static_cast<const cdb_hr::TuplePropertyType&>(i1);
        const cdb_hr::TuplePropertyType& q = static_cast<const cdb_hr::TuplePropertyType&>(i2);
        Assert::AreEqual(p.GetItems().size(), q.GetItems().size());
        for (size_t i = 0; i < p.GetItems().size(); i++)
        {
          AssertEqual(*p.GetItems()[i], *q.GetItems()[i]);
        }

        break;
      }
      case cdb_hr::TaggedPropertyTypeHybridRowSerializer::Id.Id():
      {
        const cdb_hr::TaggedPropertyType& p = static_cast<const cdb_hr::TaggedPropertyType&>(i1);
        const cdb_hr::TaggedPropertyType& q = static_cast<const cdb_hr::TaggedPropertyType&>(i2);
        Assert::AreEqual(p.GetItems().size(), q.GetItems().size());
        for (size_t i = 0; i < p.GetItems().size(); i++)
        {
          AssertEqual(*p.GetItems()[i], *q.GetItems()[i]);
        }

        break;
      }
      case cdb_hr::UdtPropertyTypeHybridRowSerializer::Id.Id():
      {
        const cdb_hr::UdtPropertyType& p = static_cast<const cdb_hr::UdtPropertyType&>(i1);
        const cdb_hr::UdtPropertyType& q = static_cast<const cdb_hr::UdtPropertyType&>(i2);
        Assert::AreEqual(p.GetName(), q.GetName());
        Assert::AreEqual(p.GetSchemaId(), q.GetSchemaId());
        break;
      }
      default:
        Assert::Fail(L"Type is abstract.");
      }
    }

    static void AssertEqual(const cdb_hr::EnumSchema& s1, const cdb_hr::EnumSchema& s2)
    {
      Assert::AreEqual(s1.GetType(), s2.GetType());
      Assert::AreEqual(s1.GetApiType(), s2.GetApiType());
      Assert::AreEqual(s1.GetName(), s2.GetName());
      Assert::AreEqual(s1.GetComment(), s2.GetComment());

      Assert::AreEqual(s1.GetValues().size(), s2.GetValues().size());
      for (size_t i = 0; i < s1.GetValues().size(); i++)
      {
        AssertEqual(*s1.GetValues()[i], *s2.GetValues()[i]);
      }
    }

    static void AssertEqual(const cdb_hr::EnumValue& s1, const cdb_hr::EnumValue& s2)
    {
      Assert::AreEqual(s1.GetValue(), s2.GetValue());
      Assert::AreEqual(s1.GetName(), s2.GetName());
      Assert::AreEqual(s1.GetComment(), s2.GetComment());
    }

    static void AssertEqual(const cdb_hr::SchemaOptions& s1, const cdb_hr::SchemaOptions& s2)
    {
      Assert::AreEqual(s1.GetDisallowUnschematized(), s2.GetDisallowUnschematized());
      Assert::AreEqual(s1.GetEnablePropertyLevelTimestamp(), s2.GetEnablePropertyLevelTimestamp());
      Assert::AreEqual(s1.GetDisableSystemPrefix(), s2.GetDisableSystemPrefix());
      Assert::AreEqual(s1.GetAbstract(), s2.GetAbstract());
    }

    static void AssertEqual(const cdb_hr::Schema& s1, const cdb_hr::Schema& s2)
    {
      Assert::AreEqual(s1.GetVersion(), s2.GetVersion());
      Assert::AreEqual(s1.GetType(), s2.GetType());
      Assert::AreEqual(s1.GetSchemaId(), s2.GetSchemaId());
      Assert::AreEqual(s1.GetName(), s2.GetName());
      Assert::AreEqual(s1.GetComment(), s2.GetComment());

      if (!s1.GetOptions().has_value())
      {
        Assert::IsFalse(s2.GetOptions().has_value());
      }
      else
      {
        Assert::IsTrue(s2.GetOptions().has_value());
        AssertEqual(s1.GetOptions().value(), s2.GetOptions().value());
      }

      Assert::AreEqual(s1.GetPartitionKeys().size(), s2.GetPartitionKeys().size());
      for (size_t i = 0; i < s1.GetPartitionKeys().size(); i++)
      {
        AssertEqual(*s1.GetPartitionKeys()[i], *s2.GetPartitionKeys()[i]);
      }

      Assert::AreEqual(s1.GetPrimaryKeys().size(), s2.GetPrimaryKeys().size());
      for (size_t i = 0; i < s1.GetPrimaryKeys().size(); i++)
      {
        AssertEqual(*s1.GetPrimaryKeys()[i], *s2.GetPrimaryKeys()[i]);
      }

      Assert::AreEqual(s1.GetStaticKeys().size(), s2.GetStaticKeys().size());
      for (size_t i = 0; i < s1.GetStaticKeys().size(); i++)
      {
        AssertEqual(*s1.GetStaticKeys()[i], *s2.GetStaticKeys()[i]);
      }

      Assert::AreEqual(s1.GetProperties().size(), s2.GetProperties().size());
      for (size_t i = 0; i < s1.GetProperties().size(); i++)
      {
        AssertEqual(*s1.GetProperties()[i], *s2.GetProperties()[i]);
      }
    }

    static void SerializerRoundtripNamespace(const cdb_hr::Namespace& ns1)
    {
      const cdb_hr::LayoutResolver& resolver = cdb_hr::SystemSchemaLiteral::GetLayoutResolver();
      const cdb_hr::Layout& layout = resolver.Resolve(cdb_hr::NamespaceHybridRowSerializer::Id);

      cdb_hr::MemorySpanResizer<byte> resizer{0};
      cdb_hr::RowBuffer row{0, &resizer};
      row.InitLayout(cdb_hr::HybridRowVersion::V1, layout, &resolver);

      // Write the whole namespace to a row.
      cdb_hr::Result r = ns1.Write(row);
      ResultAssert::IsSuccess(r);

      // Read the namespace back.
      std::unique_ptr<cdb_hr::Namespace> pns2;
      std::tie(r, pns2) = cdb_hr::Namespace::Read(row);
      ResultAssert::IsSuccess(r);
      const cdb_hr::Namespace& ns2 = *pns2;

      // Compare the materialized row with the original in-memory object model.
      Assert::AreEqual(ns1.GetVersion(), ns2.GetVersion());
      Assert::AreEqual(ns1.GetName(), ns2.GetName());
      Assert::AreEqual(ns1.GetComment(), ns2.GetComment());
      Assert::AreEqual(ns1.GetSchemas().size(), ns2.GetSchemas().size());
      for (size_t i = 0; i < ns1.GetSchemas().size(); i++)
      {
        const cdb_hr::Schema& s1 = *ns1.GetSchemas()[i];
        const cdb_hr::Schema& s2 = *ns2.GetSchemas()[i];
        AssertEqual(s1, s2);
      }
      Assert::AreEqual(ns1.GetEnums().size(), ns2.GetEnums().size());
      for (size_t i = 0; i < ns1.GetEnums().size(); i++)
      {
        const cdb_hr::EnumSchema& s1 = *ns1.GetEnums()[i];
        const cdb_hr::EnumSchema& s2 = *ns2.GetEnums()[i];
        AssertEqual(s1, s2);
      }
    }

  TEST_METHOD_WITH_OWNER(SerializeSystemNamespaceTest, "jthunter")
    {
      const cdb_hr::Namespace& ns1 = cdb_hr::SystemSchemaLiteral::GetNamespace();
      SystemSchemaUnitTests::SerializerRoundtripNamespace(ns1);
    }
  };
}
